commitrmnocb/outcome.go

```diff
@@ -167,31 +194,123 @@ func getConsensusObservation(
 	lggr logger.Logger,
 	F int,
 	destChain cciptypes.ChainSelector,
+	offchainCfg pluginconfig.CommitOffchainConfig,
+	lastPricesUpdate time.Time,
 	aos []types.AttributedObservation,
 ) (ConsensusObservation, error) {
 	aggObs := aggregateObservations(aos)
 	fChains := fChainConsensus(lggr, F, aggObs.FChain)
+	timestampConsensus := MedianTimestamp(aggObs.Timestamps)
 
 	fDestChain, exists := fChains[destChain]
 	if !exists {
 		return ConsensusObservation{},
 			fmt.Errorf("no consensus value for fDestChain, destChain: %d", destChain)
 	}
 
+	feedPrices := feedPricesConsensus(lggr, aggObs.FeedTokenPrices, fDestChain)
+	registryPrices := registryPricesConsensus(lggr, aggObs.PriceRegistryTokenPrices, fDestChain)
+	tokenPrices := selectTokens(
+		lggr,
+		feedPrices,
+		registryPrices,
+		timestampConsensus,
+		lastPricesUpdate,
+		offchainCfg.TokenPriceBatchWriteFrequency,
+		offchainCfg.TokenInfo,
+	)
 	consensusObs := ConsensusObservation{
 		MerkleRoots: merkleRootConsensus(lggr, aggObs.MerkleRoots, fChains),
 		// TODO: use consensus of observed gas prices
-		GasPrices: make(map[cciptypes.ChainSelector]cciptypes.BigInt),
-		// TODO: use consensus of observed token prices
-		TokenPrices:        make(map[types.Account]cciptypes.BigInt),
+		GasPrices:          make(map[cciptypes.ChainSelector]cciptypes.BigInt),
+		TokenPrices:        tokenPrices,
 		OnRampMaxSeqNums:   onRampMaxSeqNumsConsensus(lggr, aggObs.OnRampMaxSeqNums, fChains),
 		OffRampNextSeqNums: offRampMaxSeqNumsConsensus(lggr, aggObs.OffRampNextSeqNums, fDestChain),
 		FChain:             fChains,
+		Timestamp:          timestampConsensus,
 	}
 
 	return consensusObs, nil
 }
 
+// Checks which tokens need to be updated based on the observed token prices and the price registry updates
+// if time passed since the last update is greater than the stale threshold, update all tokens
+// otherwise calculate deviation between the price registry and feed and include deviated token prices
+func selectTokens(
+	lggr logger.Logger,
+	medianizedFeedTokenPrices map[types.Account]cciptypes.BigInt,
+	medianizedRegistryPrices map[types.Account]cciptypes.BigInt,
+	consensusTimestamp time.Time,
+	lastPricesUpdate time.Time,
+	updateFrequency commonconfig.Duration,
+	tokenInfo map[types.Account]pluginconfig.TokenInfo,
+) map[types.Account]cciptypes.BigInt {
+	tokenPrices := make(map[types.Account]cciptypes.BigInt)
+	// if the time since the last update is greater than the update frequency, update all tokens
+	if consensusTimestamp.Sub(lastPricesUpdate) > updateFrequency.Duration() {
```

Comment:
nit: use `time.After`
```suggestion
	nextPricesUpdate := lastPricesUpdate.Add(updateFrequency.Duration())
	if consensusTimestamp.After(nextPricesUpdate) {
```